### ISSUE TITLE
Can multiple roles per user be supported?

### ISSUE THREAD

**Issue by judgej** [laravel-enso/Enso#6]
A user can often be in many roles, a customer and a supplier, for example. Is this something that can or will be supported? For us, it is a deal breaker with so many real world use-cases.

**Comment by aocneanu** [comment]
Hi,

TL;DR: Not right now.

Boring stuff:

In Enso every role is a collection of permissions, and every action that a user can make in the app has a permission associated.

Right now the role-permission system does not allow multiple roles per user. 

The reason for this is that we try to focus on developing a system that suits business apps with hundreds of users.We found that its preferable to configure permissions for clusters of users mapped on the business logic, and have tens of roles, instead of having fewer roles and managing every user with its attached ones. It also an easier structure to maintain in the long term.

For clusters of users we have Owners in Enso. Owners can be departments, and can have a set of roles.

For instance, you can have Owners like "Sales", "Procurement", "Logistics", "HR". For each Owner you can design a set of roles. For the "Sales" owner you can have something like:

- Salesman
- Local Sales Coordinator
- Regional Sales Coordinator
- Sales Manager

This way you can divide 50 people from a department under 4 roles, roles that are easy to manage and update, especially when the app is still growing.

Managing the [Role-Permissions](https://github.com/laravel-enso/RoleManager) structure is easy to deal with thanks to the RoleConfigurator component that allows you to assign everything from the interface. You can read more in the [docs](docs.laravel-enso.com).

One last thing, for your example it might be preferable to build three roles instead of two, and that roles would be *client*, *supplier*, *client-supplier*.

Cheers.

**Comment by judgej** [comment]
Thanks for the detailed explanation. So if I understand correctly, instead of just one administrator being given access to set ALL roles and permissions on the entire system, the administrator can hand subsets of the permissions off to sub-administrators to manage their own group of users, and be limited in what permissions they can give to those users.

**Comment by aocneanu** [comment]
That could be a good idea, but unfortunately that's not implemented. Usually the Sales Manager and the Regional Sales Manager, from the example above, do not have the necessary technical knowledge to manage roles themselves. This needs a deeper understanding of what permissions are and how they work. We think that configuring roles is something delicate and should be the developer's responsibility.

In our case wee are building the roles for our customers based on their specifications, and after that, the department managers can test everything in their responsibity by impersonating users with lower roles and checking what those users can see and can do in the app.

Our recommendation is that you keep the System menu for the developer and invoice more ;)
